fuzz/fuzz_targets/fuzz.rs

```diff
@@ -0,0 +1,320 @@
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        Debug::fmt(&self.ray(), f)
+    }
+}
+
+impl<const D: usize> ArbitraryRay<D> {
+    /// Produces the corresponding ray from the input.
+    fn ray(&self) -> Ray<Float, D> {
+        // Double normalize helps when the first one encounters precision issues.
+        let mut direction = (self.destination.point() - self.origin.point())
+            .normalize()
+            .normalize();
+        // Ensure no degenerate direction.
+        if direction.magnitude() < 0.5 || direction.iter().any(|f| f.is_nan() || f.abs() > 1.5) {
+            direction.iter_mut().for_each(|f| *f = 1.0);
+            direction = direction.normalize();
+        }
+        assert!(
+            direction.magnitude() - 1.0 < 0.1,
+            "{}",
+            direction.magnitude()
+        );
+        let mut ray = Ray::new(self.origin.point(), direction);
+
+        if self.mode.is_grid() {
+            let max_axis = ray
+                .direction
+                .iter()
+                .enumerate()
+                .max_by(|(_, a), (_, b)| a.abs().partial_cmp(&b.abs()).unwrap())
+                .map(|(i, f)| (i, *f))
+                .unwrap();
+
+            ray.origin.iter_mut().for_each(|f| *f = f.round());
+
+            // The resulting ray will be parallel to an axis.
+            ray.direction.iter_mut().enumerate().for_each(|(i, f)| {
+                *f = if i == max_axis.0 {
+                    1f32.copysign(max_axis.1)
+                } else {
+                    0.0
+                }
+            });
+        }
+
+        ray
+    }
+}
+
+/// An arbitrary mutation to apply to the BVH to fuzz BVH optimization.
+#[derive(Debug, Arbitrary)]
+enum ArbitraryMutation<const D: usize> {
+    Remove(usize),
+    Add(ArbitraryShape<D>),
+}
+
+#[derive(Copy, Clone, Debug, Arbitrary)]
+/// Controls whether the input is modified to help test certain properties.
+enum Mode {
+    /// AABB's may have mostly arbitrary bounds, and ray may have mostly arbitrary
+    /// origin and direction.
+    Chaos,
+    /// AABB's are unit cubes, and must have integer coordinates. Ray must have an
+    /// origin consisting of integer coordinates and a direction that is parallel to
+    /// one of the axes.
+    ///
+    /// In this mode, all types of traversal are expected to yield the same results,
+    /// except when bugs exist that have yet to be fixed.
+    Grid,
+}
+
+impl Mode {
+    fn is_grid(self) -> bool {
+        matches!(self, Self::Grid)
+    }
+}
+
+/// The complete set of inputs for a single fuzz iteration.
+#[derive(Debug, Arbitrary)]
+struct Workload<const D: usize> {
+    shapes: Vec<ArbitraryShape<D>>,
+    ray: ArbitraryRay<D>,
+    mutations: Vec<ArbitraryMutation<D>>,
+}
+
+impl<const D: usize> Workload<D> {
+    /// Called directly from the `cargo fuzz` entry point. Code in this function is
+    /// easier for `rust-analyzer`` than code in that macro.
+    fn fuzz(mut self) {
```

Comment:
Perhaps make the reader aware to look at the module level documentation to get the gist of what we're trying to do in this function.